python/dgl/data/explain_syn.py

```diff
@@ -0,0 +1,757 @@
+    - Perturb the graph by adding random edges.
+    - Generate constant feature for all nodes, which is 1.
+    - Nodes in the tree belong to class 0 and nodes in grids belong to class 1.
+
+    Parameters
+    ----------
+    tree_height : int, optional
+        Height of the balanced binary tree. Default: 8
+    num_motifs : int, optional
+        Number of grid motifs to use. Default: 80
+    perturb_ratio : float, optional
+        Number of random edges to add in perturbation divided by the
+        number of original edges in the graph. Default: 0.1
+    seed : integer, random_state, or None, optional
+        Indicator of random number generation state. Default: None
+    raw_dir : str, optional
+        Raw file directory to store the processed data. Default: ~/.dgl/
+    force_reload : bool, optional
+        Whether to always generate the data from scratch rather than load a cached version.
+        Default: False
+    verbose : bool, optional
+        Whether to print progress information. Default: True
+    transform : callable, optional
+        A transform that takes in a :class:`~dgl.DGLGraph` object and returns
+        a transformed version. The :class:`~dgl.DGLGraph` object will be
+        transformed before every access. Default: None
+
+    Attributes
+    ----------
+    num_classes : int
+        Number of node classes
+
+    Examples
+    --------
+
+    >>> from dgl.data import TreeGrids
+    >>> dataset = TreeGrids()
+    >>> dataset.num_classes
+    2
+    >>> g = dataset[0]
+    >>> label = g.ndata['label']
+    >>> feat = g.ndata['feat']
+    """
+    def __init__(self,
+                 tree_height=8,
+                 num_motifs=80,
+                 perturb_ratio=0.1,
+                 seed=None,
+                 raw_dir=None,
+                 force_reload=False,
+                 verbose=True,
+                 transform=None):
+        self.tree_height = tree_height
+        self.num_motifs = num_motifs
+        self.perturb_ratio = perturb_ratio
+        self.seed = seed
+        super(TreeGrids, self).__init__(name='TREE-GRIDS',
+                                        url=None,
+                                        raw_dir=raw_dir,
+                                        force_reload=force_reload,
+                                        verbose=verbose,
+                                        transform=transform)
+
+    def process(self):
+        if self.seed is not None:
+            np.random.seed(self.seed)
+
+        g = nx.balanced_tree(r=2, h=self.tree_height)
+        edges = list(g.edges())
+        src, dst = map(list, zip(*edges))
+        n = nx.number_of_nodes(g)
+
+        # Nodes in the base tree graph belong to class 0
+        node_labels = [0] * n
+        # The motifs will be evenly attached to the nodes in the base graph.
+        spacing = math.floor(n / self.num_motifs)
+
+        # Construct a three-by-three grid
+        motif_g = nx.grid_graph([3, 3])
```

Comment:
Consider expose number of grid graph nodes to argument.

Reply:
I also suggest to add some graph transform utility such as `add_ring_motif`, `add_house_motif`, etc.

Reply:
Done for exposing an argument controlling the number of grid graph nodes

Reply:
I'm not sure about the benefit of having utility functions like `add_ring_motif` and `add_house_motif`. I've only seen the use of them in constructing these synthetic datasets for GNN explainability. Also it's not clear how to attach motifs to a graph in a general way beyond these datasets.

Reply:
I see. That makes sense.